### ISSUE TITLE
Question: How to generate a swagger spec

### ISSUE THREAD

**Issue by TFTomSun** [RSuter/NSwag#1661]
What is the best way to get started if I want to generate a swagger spec (SwaggerDocument) including operations, components etc. 
I already started with that snippet:

`  var document = new SwaggerDocument();
            document.Info.Version = "1.0";
            document.OpenApi = "3.0.0";
            document.Info.Title = "TestSwagger";
            var pathItem = new SwaggerPathItem();
            var operation = new SwaggerOperation();
            var response = new SwaggerResponse();
            response.Content.Add("TestResponseValue", new OpenApiMediaType()
            {
                 Schema = new NJsonSchema.JsonSchema4 {  Type = NJsonSchema.JsonObjectType.e}
            })
            operation.Responses.Add("TestResponse",response);
            pathItem.Add(SwaggerOperationMethod.Get, operation);

            document.Paths.Add("TestPath",new SwaggerPathItem() {  })
`
How can I add components and use them in operations as parameters or responses?

**Comment by zuckerthoben** [comment]
Do you really want to generate all this "by hand" or do you want to programmatically generate the SwaggerDocument based on e.g. a Web Api? 
It never occured to me that you would write all this yourself if you use NSwag (yet).

**Comment by TFTomSun** [comment]
yes. i want that. I don't have a static service. I want to provide a dynamic service and generate a client for it.

**Comment by RSuter** [comment]
If you want to completely build up the spec then this is the way to go... For components, you'd add a schema or other object to the document.Components.Xzy collection and then somewhere else reference it with `Schema = new JsonSchema4 { Reference = mySchema }` (mySchema must be added to components). When you call `document.ToJson()` the serializer will generate $refs...

**Comment by TFTomSun** [comment]
@RSuter thanks. Let's assume I have some predefined classes for which I want to generate schema's. When i use JsonSchema4.FromTypeAsync i just get back one schema. But I would like to have all subclasses / enums as seperate schemas.

for example:
```csharp
  public enum TestEnum
        {
            A,
            B,
            C,
        }

        public enum ConstantEnum
        {
            ConstantValue
        }

        class TestChildClass
        {
            public ConstantEnum TestConstantProperty { get; set; }

            public TestEnum TestEnumProperty { get; set; }

        }

        class TestClass
        {
            public TestChildClass Child { get; set; }
            public ConstantEnum TestConstantProperty { get; set; }

            public TestEnum TestEnumProperty { get; set; }
        }

```
When I generate a schema for the root class TestClass, it will be converted to one schema which ends up as one component in the swagger spec. But I would prefer to get one schema per type for TestClass, TestChildClass, TestEnum and ConstantEnum. And they should just have references on each other. Is there a possibility to achieve that?

**Comment by RSuter** [comment]
This is more advanced and you need to avoid the static method fromtypeasync and instead use a schema generator and reuse a schema resolver

**Comment by TFTomSun** [comment]
Hm ... isn't it generated like that when I generate the swagger spec for a asp.net core service implementation? I expected that to be the default, because the types can be multiple reused else where. Can you point me to some implementation where this is done?

**Comment by RSuter** [comment]
See 
https://github.com/RSuter/NSwag/blob/master/src/NSwag.SwaggerGeneration.AspNetCore/AspNetCoreToSwaggerGenerator.cs

Or simplified but not swagger: 
https://github.com/RSuter/SigSpec/blob/master/src/SigSpec.Core/SigSpecGenerator.cs#L24

**Comment by RSuter** [comment]
> Hm ... isn't it generated like that when I generate the swagger spec for a asp.net core service implementation? 

Yes, it is. See my links. I assumed that you want to implement a new generator like the asp.net core one but from scratch?

**Comment by TFTomSun** [comment]
Well actually I would prefer not to implement a new one. But yes, probably, since I don't have a static Asp.NET core service implementation. I have an existing library (with a deep big object model) and want to provide a technology independend (rest) access to it. Implementing a Controller layer on top of it isn't feasible because it would mean to implement thousand of methods that just point to somewhere in the object model. Thats why I thought to do it dynamically on the server side, and generate a swagger spec based on that object model, that can be used for generating a clients.

**Comment by RSuter** [comment]
I think it would be easier to implement a custom document processor which adds the dynamic operations and schemas to the pregenerated (empty) document: 
https://github.com/RSuter/NSwag/wiki/Document-Processors-and-Operation-Processors

The resolver can be accessed via context parameter (use it to generate referenced schemas).

**Comment by TFTomSun** [comment]
So what I basically need is a swagger spec that result in the same object model on client side with some ids that help me to identify the same objects on server side
